records/2023-01-26.bounce-rate-automated-emergency-brake.md

```diff
@@ -0,0 +1,83 @@
+
+### Provide bounce rate guidance
+Provide guidance to service owners to explain bounce rate concepts and how they apply to their service.  Provide guidance on how services can keep their bounce rate low. 
+
+### bounce-rate-threshold calculation options
+When calculating a service's bounce rate, we can either consider them in isolation or in aggregate with other services.  The bounce rate is calculated as a percentage of the total number of emails sent.  If we consider a service in isolation, then the bounce rate will be calculated as a percentage of the total number of emails sent by that service.  If we consider a service in aggregate, then the bounce rate will be calculated as a percentage of the total number of emails sent by all services.
+
+#### Option 1: Consider a service in isolation
+Considering a service in isolation would mean calculating the bounce rate solely based on an individual services activity.  The benefit of this is that the calculation would be straight-forward and easy to understand by the service.  The downside is that even if a particular service has a high bounce rate, it may not have a large impact on the overall bounce rate. 
+
+#### Option 2: Consider a service in aggregate
+Considering a service in aggregate would mean calculating the bounce rate based on the total number of emails sent by all services.  The benefit of this is that the calculation would be more accurate and would reflect the impact of a particular service on the overall bounce rate.  The downside here is that the calculation would be more complex and may be difficult to understand by the service, since for example even if they had an extremely high bounce rate, we might not indicate it as problematic if it has little impact on the overall bounce rate.
+
+## Staggered roll-out
+To prevent suspending services that may not actually be causing issues with the overall bounce rate, it is suggested to roll out this feature in stages.  The first stage would include tracking bounces, calculating bounce rates, and displaying them on service dashboards. It would also include bounce rate guidance for services. Once we are sure our calculations work satisfactorily with live data, we would move to the second stage which will include notifying services when they are approaching the bounce rate threshold and suspending services when they reach the threshold.
+
+## Additional considerations
+
+### How GC Notify handles bounce notifications from SES
+GC Notify uses a lambda function named `ses-to-sqs-email-callbacks` to handle bounce notifications sent back from SES.  When a callback comes in, the lambda function parses the bounce notification and sends it a queue to be picked up and run by the celery task `process-ses-result`.  The `process-ses-result` task updates the status of the notification in the database including whether it was a hard or a soft bounce.
```

Comment:
This explains well how the signal for bounce notification might come into the Celery task. We could also explain a) how we plan to record it here, if we should leverage Redis or other tool, b) what's the time period we should use, c) how do we map the service to the received bounce message.

Reply:
Added extra details to ADR.